pennylane/ops/functions/assert_valid.py

```diff
@@ -57,14 +57,27 @@ def _check_decomposition(op, skip_wire_mapping):
 
         assert isinstance(decomp, list), "decomposition must be a list"
         assert isinstance(compute_decomp, list), "decomposition must be a list"
-        assert op.__class__ not in [
-            decomp_op.__class__ for decomp_op in decomp
-        ], "an operator should not be included in its own decomposition"
+        assert op not in decomp, "an operator should not be included in its own decomposition"
 
         for o1, o2, o3 in zip(decomp, compute_decomp, processed_queue):
             assert o1 == o2, "decomposition must match compute_decomposition"
             assert o1 == o3, "decomposition must match queued operations"
             assert isinstance(o1, qml.operation.Operator), "decomposition must contain operators"
+            if not decomposition_to_same_class:
```

Comment:
Personally I feel that checking whether an operator decomposes to exactly itself is sufficient.

Reply:
I'd also be happy with this, let me check in at the standup tomorrow and make sure no one has objections